File: mobile/src/main/java/net/mm2d/dmsexplorer/domain/entity/ContentDirectoryEntity.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.domain.entity;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import net.mm2d.android.upnp.cds.CdsObject;
import net.mm2d.dmsexplorer.domain.model.ExploreListener;
import net.mm2d.dmsexplorer.domain.model.ExploreListenerAdapter;
import net.mm2d.util.Log;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class ContentDirectoryEntity implements DirectoryEntity {
    private static final String ROOT_OBJECT_ID = "0";
    private static final String ROOT_TITLE = "";
    @NonNull
    private final String mParentId;
    @NonNull
    private final String mParentTitle;
    @Nullable
    private ContentEntity mSelectedEntity;
    @NonNull
    private final List<ContentEntity> mList = new ArrayList<>();
    private static final ExploreListener ENTRY_LISTENER = new ExploreListenerAdapter();
    @NonNull
    private ExploreListener mEntryListener = ENTRY_LISTENER;
    private volatile boolean mInProgress = true;
    private volatile Disposable mDisposable;

    public ContentDirectoryEntity() {
        this(ROOT_OBJECT_ID, ROOT_TITLE);
    }

    private ContentDirectoryEntity(
            @NonNull final String parentId,
            @NonNull final String parentTitle) {
        mParentId = parentId;
        mParentTitle = parentTitle;
    }

    public void terminate() {
        setExploreListener(null);
        if (mDisposable != null) {
            mDisposable.dispose();
            mDisposable = null;
        }
    }

    @NonNull
    public String getParentId() {
        return mParentId;
    }

    @NonNull
    @Override
    public String getParentName() {
        return mParentTitle;
    }

    @Override
    public boolean isInProgress() {
        return mInProgress;
    }

    @NonNull
    @Override
    public List<ContentEntity> getEntities() {
        return mList;
    }

    @Nullable
    public ContentDirectoryEntity enterChild(@NonNull final ContentEntity entity) {
        if (!mList.contains(entity)) {
            return null;
        }
        if (entity.getType() != ContentType.CONTAINER) {
            return null;
        }
        mSelectedEntity = entity;
        final CdsObject object = (CdsObject) entity.getObject();
        return new ContentDirectoryEntity(object.getObjectId(), object.getTitle());
    }

    @Override
    public void setSelectedEntity(@Nullable final ContentEntity entity) {
        if (!mList.contains(entity)) {
            mSelectedEntity = null;
            return;
        }
        mSelectedEntity = entity;
    }

    @Nullable
    @Override
    public ContentEntity getSelectedEntity() {
        return mSelectedEntity;
    }

    public void setExploreListener(@Nullable final ExploreListener listener) {
        mEntryListener = listener != null ? listener : ENTRY_LISTENER;
    }

    public void clearState() {
        if (mDisposable != null) {
            mDisposable.dispose();
            mDisposable = null;
        }
        mSelectedEntity = null;
        mInProgress = true;
        mList.clear();
        mEntryListener.onStart();
    }

    public void startBrowse(@NonNull final Observable<CdsObject> observable) {
        mEntryListener.onStart();
        mDisposable = observable
                .subscribe(object -> {
                    mList.add(new CdsContentEntity(object));
                    mEntryListener.onUpdate(mList);
                }, Log::w, () -> {
                    mInProgress = false;
                    mEntryListener.onComplete();
                });
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/domain/entity/ContentDirectoryEntity.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.domain.entity

import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import net.mm2d.android.upnp.cds.CdsObject
import net.mm2d.dmsexplorer.domain.model.ExploreListener
import net.mm2d.dmsexplorer.domain.model.ExploreListenerAdapter
import net.mm2d.dmsexplorer.settings.Settings
import net.mm2d.dmsexplorer.settings.SortKey.*
import net.mm2d.log.Logger

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class ContentDirectoryEntity private constructor(
    val parentId: String,
    override val parentName: String
) : DirectoryEntity {
    private val list = mutableListOf<ContentEntity>()
    private var sortedList = emptyList<ContentEntity>()
    private var entryListener: ExploreListener = ENTRY_LISTENER
    @Volatile
    override var isInProgress = true
        private set
    @Volatile
    private var disposable: Disposable? = null

    override val entities: List<ContentEntity>
        get() = sortedList

    private var _selectedEntity: ContentEntity? = null
    override var selectedEntity: ContentEntity?
        get() = _selectedEntity
        set(entity) {
            if (entity == null || !list.contains(entity)) {
                _selectedEntity = null
                return
            }
            _selectedEntity = entity
        }

    constructor() : this(ROOT_OBJECT_ID, ROOT_TITLE)

    fun terminate() {
        setExploreListener(null)
        disposable?.dispose()
        disposable = null
    }

    fun enterChild(entity: ContentEntity): ContentDirectoryEntity? {
        if (!list.contains(entity)) {
            return null
        }
        if (entity.type != ContentType.CONTAINER) {
            return null
        }
        _selectedEntity = entity
        val cdsObject = entity.cdsObject as CdsObject
        return ContentDirectoryEntity(cdsObject.objectId, cdsObject.title)
    }

    fun setExploreListener(listener: ExploreListener?) {
        entryListener = listener ?: ENTRY_LISTENER
    }

    fun clearState() {
        disposable?.dispose()
        disposable = null
        _selectedEntity = null
        isInProgress = true
        list.clear()
        sortedList = emptyList()
        entryListener.onStart()
    }

    fun startBrowse(observable: Observable<CdsObject>) {
        entryListener.onStart()
        disposable = observable
            .subscribeOn(Schedulers.io())
            .subscribe({ cdsObject ->
                list.add(CdsContentEntity(cdsObject))
                sort()
                entryListener.onUpdate(sortedList)
            }, {
                Logger.w(it)
            }, {
                isInProgress = false
                entryListener.onComplete()
            })
    }

    private fun sort() {
        val settings = Settings.get()
        val isAscending = settings.isAscendingSortOrder
        sortedList = when (settings.sortKey) {
            NONE -> list.toList()
            NAME -> {
                if (isAscending) list.sortedBy { it.name }
                else list.sortedByDescending { it.name }
            }
            DATE -> {
                if (isAscending) list.sortedBy { it.date }
                else list.sortedByDescending { it.date }
            }
        }
    }

    fun onUpdateSortSettings() {
        Completable.fromAction {
            sort()
            entryListener.onUpdate(sortedList)
        }
            .subscribeOn(Schedulers.io())
            .subscribe()
    }

    companion object {
        private const val ROOT_OBJECT_ID = "0"
        private const val ROOT_TITLE = ""
        private val ENTRY_LISTENER: ExploreListener = ExploreListenerAdapter()
    }
}
